### PR TITLE
fix(Core/Misc): all GCC warnings

### PR DESCRIPTION
**Pull request by FrancescoBorzi**
## CHANGES PROPOSED:
-  Fixed most of GCC warnings
-  Disabled a couple of GCC warnings in order to have a clean build, which will allow adding GCC to our CI. The goal is merely to prevent more warnings to get to master. I will open issues for those ignored warnings so we can fix them later.

## TESTS PERFORMED:
- None

## HOW TO TEST THE CHANGES:

Theoretically, this PR should not introduce any change in the game, however it's always better to double check, so please:

- checkout this PR and play the game for a while, make sure there is nothing broken by this change

I know it's too generic, but I can't provide more detailed test instructions this time.

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers can anyone please give this a try?

**Comment by peti446** [comment]
Build it on linux with GCC, no problems building it, server starts with no errors and runs for more than 10 minutes.
In terms of gameplay, nothing is apparently broken, used AH, mail, killed, talked to npc, LFG.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -8593,6 +8596,7 @@ bool Unit::HandleProcTriggerSpell(Unit* victim, uint32 damage, AuraEffect* trigg
 
             target = this;
             trigger_spell_id = 22588;
+            [[fallthrough]]; // TODO: Not sure whether the fallthrough was a mistake (forgetting a break) or intended. This should be double-checked.
```
Comment (Viste) [inline src/server/game/Entities/Unit/Unit.cpp:8599]: @FrancescoBorzi  forgetting break proofs
https://github.com/SkittlesAreFalling/WoW-Carnage/blob/45d5d483c857bd21135f70e4abf278175ddf40e6/src/game/UnitAuraProcHandler.cpp#L905
https://github.com/Kittnz/ZeroServer/blob/193794dcdbda45a5bd512e6d83780e42308734e5/src/game/UnitAuraProcHandler.cpp#L905
https://github.com/IldebrandoS/brandonie-craft/blob/d36bf42f4ed163c821591c9d94539bd77d8462a2/game/UnitAuraProcHandler.cpp#L1236

Comment (FrancescoBorzi) [reply src/server/game/Entities/Unit/Unit.cpp:8599]: thanks @Viste we can open another PR for that!

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -18981,7 +18980,7 @@ void Player::_LoadMailAsynch(PreparedQueryResult result)
                 m->receiver       = fields[17].GetUInt32();
                 m->subject        = fields[18].GetString();
                 m->body           = fields[19].GetString();
-                has_items         = fields[20].GetBool();
+//                has_items         = fields[20].GetBool();
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.cpp:18983]: rofl

File: src/server/scripts/Northrend/FrozenHalls/PitOfSaron/pit_of_saron.cpp [hunk 3]
```diff
@@ -421,8 +425,8 @@ class npc_pos_deathwhisper_necrolyte : public CreatureScript
         void EnterCombat(Unit* /*who*/)
         {
             // TODO This spell check is invalid
-            if (SPELL_NECROLYTE_CHANNELING)
-                me->RemoveAura(SPELL_NECROLYTE_CHANNELING);
+//            if (SPELL_NECROLYTE_CHANNELING)
```
Comment (Winfidonarleyan) [inline src/server/scripts/Northrend/FrozenHalls/PitOfSaron/pit_of_saron.cpp:428]: maybe add has aura?

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (FrancescoBorzi) [reply src/server/game/Entities/Player/Player.cpp:18983]: yeah, it was unused there

File: src/server/game/Scripting/ScriptMgr.h [hunk 4]
```diff
@@ -787,6 +787,9 @@ class AchievementCriteriaScript : public ScriptObject
 
         bool IsDatabaseBound() const { return true; }
 
+// TODO: manually silencing this warning, but it should properly fixed
+#pragma GCC diagnostic ignored "-Woverloaded-virtual"
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:791]: https://github.com/azerothcore/azerothcore-wotlk/pull/3473/files

Comment (FrancescoBorzi) [reply src/server/game/Scripting/ScriptMgr.h]: @Winfidonarleyan thank you

File: src/server/game/Entities/Unit/Unit.cpp (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Unit/Unit.cpp:8599]: need break

**Review by Winfidonarleyan — APPROVED** [review]

**Pull request merged**